WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingManager.java

```diff
@@ -177,33 +219,171 @@ public void submitRequestForm(User user, Container container, WNPRC_PurchasingCo
                 List<Map<String, Object>> rowsInserted = qus.insertRows(user, container, newVendorData, errors, null, null);
 
                 //set new vendor's id
-                if (rowsInserted.size() == 1 && purchasingRequestsData.get(0).get("vendorId") == null) {
+                if (rowsInserted.size() == 1 && purchasingRequestsData.get(0).get("vendorId") == null)
                     purchasingRequestsData.get(0).put("vendorId", rowsInserted.get(0).get("rowId"));
-                }
             }
 
             if (purchasingRequestsTable != null)
             {
                 qus = purchasingRequestsTable.getUpdateService();
                 assert qus != null;
-                qus.insertRows(user, container, purchasingRequestsData, errors, null, null);
+                if (isNewRequest)
+                    insertedPurchasingReq = qus.insertRows(user, container, purchasingRequestsData, errors, null, null);
+                else
+                    insertedPurchasingReq = qus.updateRows(user, container, purchasingRequestsData, null, null, null);
+
+                if (null != insertedPurchasingReq)
+                    requestForm.setRowId((Integer) insertedPurchasingReq.get(0).get("rowId"));
+                else
+                {
+                    requestOrderErrors.addError(new SimpleValidationError("Unable to submit purchasing request"));
+                    return validationErrors;
+                }
+            }
+
+            //Line items data
+            List<Map<String, Object>> newLineItemsData = new ArrayList<>();
+            List<Map<String, Object>> updatedLineItemsData = new ArrayList<>();
+            List<Map<String, Object>> deleteLineItemsData = new ArrayList<>();
+            int index = 0;
+            for (JSONObject lineItem : requestForm.getLineItems())
+            {
+                row = new CaseInsensitiveHashMap<>();
+                row.put("requestRowId", insertedPurchasingReq.get(0).get("rowId"));
+
+                if (null == lineItem.get("item"))
+                    lineItemErrors.addError(new PropertyValidationError("Required value for 'Part no./Item description' not provided", "item"));
+                else
+                    row.put("item", lineItem.get("item"));
+
+                if (null == lineItem.get("itemUnit"))
+                    lineItemErrors.addError(new PropertyValidationError("Required value for 'Unit' not provided", "itemUnit"));
+                else
+                    row.put("itemUnitId", lineItem.get("itemUnit"));
+
+                if (null == lineItem.get("unitCost"))
+                    lineItemErrors.addError(new PropertyValidationError("Required numeric value for 'Unit Cost' not provided", "unitCost"));
```

Comment:
Are these checks needed? I think you could make sure the field's metadata knows it's required and the insertRows() and updateRows() should handle it.

Reply:
It is not needed for the required fields indeed, but i added it explicitly because the errors from insertRows() and updateRows() were not as user friendly.

Reply:
Is it mostly around wanting to use the label ("Unit Cost") instead of the column name ("unitCost")?

Reply:
Mainly because the error was "Missing value for required property: unitCost" -- thought it was odd to see unitCost as "property" in the message/didn't think it was user friendly.